#! /bin/sh /usr/share/dpatch/dpatch-run
## bts619123_mkdir_endless_loop_fix.dpatch by Jonathan Nieder
## <jrnieder@gmail.com>
##
## DP: common: check_create_dir(): Support symlinks as well.
## DP:
## DP: Previously, the following situation would cause an endless loop (as
## DP: reported by Michael Prokop in Debian bug #619123): the (CSV or RRD)
## DP: datadir is a symlink pointing to a non-existent target.
## DP:
## DP: With this patch applied, check_create_dir() fails with "<file> exists
## DP: but is not a directory".

@DPATCH@

diff a/src/common.c b/src/common.c
--- a/src/common.c
+++ b/src/common.c
@@ -542,7 +542,8 @@ int check_create_dir (const char *file_orig)
 		}
 
 		while (42) {
-			if (stat (dir, &statbuf) == -1)
+			if ((stat (dir, &statbuf) == -1)
+					&& (lstat (dir, &statbuf) == -1))
 			{
 				if (errno == ENOENT)
 				{
